cmd/snap/cmd_run.go

```diff
@@ -1013,3 +1057,170 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
+	// auxUnits []struct {
+	//   Name string
+	//   Properties []struct{
+	//   	Name string
+	//   	Value interface{}
+	//	 }
+	// } // auxUnits describe any additional units to define.
+	type property struct {
+		Name string
+		// XXX: This is getting marshaled as an invalid variant.
+		Value interface{}
+	}
+	type auxUnit struct {
+		Name  string
+		Props []property
+	}
+
+	mode := "fail"
+	properties := []property{{"PIDs", []uint{uint(pid)}}}
+	aux := []auxUnit(nil)
+	systemd := conn.Object("org.freedesktop.systemd1", "/org/freedesktop/systemd1")
+	call := systemd.Call(
+		"org.freedesktop.systemd1.Manager.StartTransientUnit",
+		0, /* call flags */
+		unitName,
+		mode,
+		properties,
+		aux,
+	)
+	var job dbus.ObjectPath
+	if err := call.Store(&job); err != nil {
+		if dbusErr, ok := err.(dbus.Error); ok {
+			switch dbusErr.Name {
+			case "org.freedesktop.DBus.Error.UnknownMethod":
+				// The DBus API is not supported on this system. This can happen on
+				// very old versions of Systemd, for instance on Ubuntu 14.04.
+				logger.Debugf("cannot create transient scope on this system: %s", err)
+				return nil
+			case "org.freedesktop.systemd1.UnitExists":
+				// Starting a scope with a name that already exists is an
+				// error. Normally this should never happen.
+				return fmt.Errorf("cannot create transient scope: scope %q clashed: %s", unitName, err)
+			case "org.freedesktop.DBus.Error.Spawn.ChildExited":
+				// The system does not have a DBus session bus.
+				// Consume the error and continue with extra checks below.
+				logger.Debugf("session bus is not available, consider installing dbus-user-session")
+				err = nil
+			}
+		}
+		if err != nil {
+			return fmt.Errorf("cannot create transient scope: %s", err)
+		}
+	}
+	return nil
+}
+
+var dbusSystemBus = func() (*dbus.Conn, error) {
+	return dbus.SystemBus()
+}
+
+var dbusSessionBus = func() (*dbus.Conn, error) {
+	return dbus.SessionBus()
+}
+
+var randomUUID = func() (string, error) {
+	// The source of the bytes generated here is the same as that of
+	// /dev/urandom which doesn't block and is sufficient for our purposes
+	// of avoiding clashing UUIDs that are needed for all of the non-service
+	// commands that are started with the help of this UUID.
+	uuidBytes, err := ioutil.ReadFile("/proc/sys/kernel/random/uuid")
+	return strings.TrimSpace(string(uuidBytes)), err
+}
+
+var osGetuid = os.Getuid
+var osGetpid = os.Getpid
+
+var cgroupProcessPathInTrackingCgroup = cgroup.ProcessPathInTrackingCgroup
+
+var createTransientScope = func(securityTag string) error {
+	if !features.RefreshAppAwareness.IsEnabled() {
+		return nil
+	}
+	logger.Debugf("creating transient scope %s", securityTag)
+	// The scope is created with a DBus call to systemd running either on
+	// system or session bus, depending on if we are starting a program as root
+	// or as a regular user.
+	var conn *dbus.Conn
+	var err error
+	if uid := osGetuid(); uid == 0 {
```

Comment:
Wondering if this check is enough. A case discussed on IRC, say I log in as a `root`, and run ohmygiraffe. Should we use the system scope or session here?

From what I could gather:
- services started by systemd do not seem to have DBUS_SESSION_BUS_ADDRESS set
- a root session has both /run/user/0/bus and DBUS_SESION_BUS_ADDRESS

Maybe we should check both the socket & env != "", or getppid() != 1?

Reply:
As discussed on IRC I changed my thinking on this and I believe we should do the following:

- non-service applications always use the session bus, including for root
- service applications with both system and user scope are not tracked with a new scope (no change here, they already have the right tracking)
- hooks always use the system bus - we need to track them and they will never run inside a session

I will make the adjustments shortly.